### PR TITLE
Refactor retry logic to handle conflict on updates.

### LINKED ISSUE #25: oathkeeper-maester crashes on updating configmap w/ rules
**Description**

Oathkeeper-maester crashes while updating configmap with rules.  Controller should have the most recent version of the configmap to update.

**Expected result**
The most recent version of configmap should be updated. Oathkeeper-maester should log the error and should NOT crash.

**Actual result**

The controller logs error on updating old version of configmap with rules and crashes. Oathkeeper-maester pod restarted several times. See the following logs:
`oathkeeper-maester-699cccb749-vplhm                                   0/1     CrashLoopBackOff   26`
```
2019-11-06T10:17:21.490Z	ERROR	controllers.Rule	unable to process rules Configmap	{"error": "All attempts fail:\n#1: Operation cannot be fulfilled on configmaps \"ory-oathkeeper-rules\": the object has been modified; please apply your changes to the latest version and try again\n#2: Operation cannot be fulfilled on configmaps \"ory-oathkeeper-rules\": the object has been modified; please apply your changes to the latest version and try again\n#3: Operation cannot be fulfilled on configmaps \"ory-oathkeeper-rules\": the object has been modified; please apply your changes to the latest version and try again\n#4: Operation cannot be fulfilled on configmaps \"ory-oathkeeper-rules\": the object has been modified; please apply your changes to the latest version and try again\n#5: Operation cannot be fulfilled on configmaps \"ory-oathkeeper-rules\": the object has been modified; please apply your changes to the latest version and try again"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
github.com/ory/oathkeeper-maester/controllers.(*RuleReconciler).Reconcile
	/go/src/github.com/ory/oathkeeper-maester/controllers/rule_controller.go:108
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.2/pkg/internal/controller/controller.go:216
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.2/pkg/internal/controller/controller.go:192
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.2.0-beta.2/pkg/internal/controller/controller.go:171
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:152
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:153
k8s.io/apimachinery/pkg/util/wait.Until
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190404173353-6a84e37a896d/pkg/util/wait/wait.go:88
```

Version of the maester installed: oryd/oathkeeper-maester:v0.0.2-beta.1

### PR DESCRIPTION
**Pull request by Tomasz-Smelcerz-SAP**
This PR implements retries on a race condition on ConfigMap updates for simulatneous changes to multiple rules.
This is probably caused by K8s apiserver delays or sync period between apiserver instances in bigger clusters, it never occurs in Minikube.
The fix is to retry the entire logic path: fetch, update/create.
If the first update fails, the retry will re-fetch the ConfigMap and the `the object has been modified; please apply your changes to the latest version and try again` condition is handled.

Fixes #25

### CODE REVIEW

**Comment by Tomasz-Smelcerz-SAP** [comment]
@aeneasr @piotrmsc CI is failing, my short investigation is: instead of `go install sigs.k8s.io/kustomize/v3/cmd/kustomize` we should do `go install sigs.k8s.io/kustomize/kustomize/v3`.

**Comment by piotrmsc** [comment]
@Tomasz-Smelcerz-SAP IMHO is not problem with customize, IMO problem is with go-client version which is used inside project, check go.mod && go.sum, maybe go mod tidy could help ?

**Comment by Tomasz-Smelcerz-SAP** [comment]
@piotrmsc kustomize command was wrong, because it's failing locally with the same error. Now everything works on my machine, but fails on CI. You're right, there's something with versions of go-client.

**Comment by Tomasz-Smelcerz-SAP** [comment]
@aeneasr It's ready for the review :)
We were struggling with the CI, but now the config is updated and simplified. Next PRs should go smoothly until another major golang upgrade.

File: .circleci/config.yml [hunk 1]
```diff
@@ -56,11 +49,10 @@ jobs:
           name: Update golang
           command: |
             sudo rm -rf /usr/local/go/
-            curl -LO https://dl.google.com/go/go1.12.7.linux-amd64.tar.gz
-            sudo tar -C /usr/local -xzf go1.12.7.linux-amd64.tar.gz
+            curl -LO https://dl.google.com/go/go1.13.3.linux-amd64.tar.gz
+            sudo tar -C /usr/local -xzf go1.13.3.linux-amd64.tar.gz
```
Comment (aeneasr) [inline .circleci/config.yml]: Why not just `circleci/golang:1.13.3`?

File: Makefile [hunk 2]
```diff
@@ -86,7 +87,7 @@ docker-push:
 controller-gen:
 ifeq (, $(shell which controller-gen))
 	go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.2.0-beta.2
-CONTROLLER_GEN=$(shell go env GOPATH)/bin/controller-gen
+CONTROLLER_GEN=$(shell which controller-gen)
```
Comment (aeneasr) [inline Makefile]: This assumes that `PATH` includes `$GOPATH/bin` which isn't always the case. Why was `$(shell go env GOPATH)/bin/` not a suitable option? `go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.2.0-beta.2` would install this in the appropriate `go env GOPATH`, right?

File: .circleci/config.yml (hunk 1 above)
Comment (Tomasz-Smelcerz-SAP) [reply .circleci/config.yml]: This step (`test-integration`) is executed not in the docker image, but on the machine itself (`machine=true`). That's why we install golang manually.

File: Makefile (hunk 2 above)
Comment (Tomasz-Smelcerz-SAP) [reply Makefile]: For `'test-integration` step, the value of `go env GOPATH` is unfortunately this: `/home/circleci/.go_workspace:/usr/local/go_workspace` - there are two folder paths separated with a colon. So the concatenation of `/bin/controller-gen` wasn't working.

Comment (aeneasr) [reply Makefile]: Oh I see - ok that makes sense now!

File: .circleci/config.yml (hunk 1 above)
Comment (aeneasr) [reply .circleci/config.yml]: Ahh, totally missed that - github concatenated the `machine: true` part :)

**Pull request merged**